lmdeploy/pytorch/engine/model_agent.py

```diff
@@ -1250,60 +1257,57 @@ def _update_is_decoding(self, forward_inputs):
         if self.cache_config.role != EngineRole.Prefill:
             self._is_decoding = not self._is_decoding
 
-    async def _broadcast_has_inputs(self, has_inputs: bool = False):
+        if self.cache_config.role == EngineRole.Decode:
+            # set timeout for next inputs
+            # next inputs is ~self._is_decoding
+            # and prefill is rarely happened in decoding engine
+            if self._is_decoding:
+                self.timeout = max(0.02, self.base_timeout / 2)
+            else:
+                self.timeout = self.base_timeout
+
+    async def _gather_has_inputs(self, has_inputs: bool = False):
         """Broadcast has inputs."""
         attn_tp_group = self.dist_ctx.attn_tp_group
         attn_tp = self.dist_ctx.dist_config.attn_tp
         if attn_tp == 1:
             return has_inputs
 
         group = attn_tp_group.cpu_group
-        rank = dist.get_global_rank(group, 0)
-        has_inputs = torch.tensor((has_inputs, ))
-        handle = dist.broadcast(has_inputs, src=rank, group=group, async_op=True)
+        has_inputs = torch.tensor((int(has_inputs), ))
```

Comment:
Should we put `has_inputs` on cuda device?

Reply:
This all reduce is used to check if inputs on same tp groups are ready or not.
move inputs on gpu might block H2D of model inputs.